### PR TITLE
Initial dataset structure

### PR DESCRIPTION
**Pull request by wiederm**
## Description
Outline the basic dataset structure

This PR contains the structure for a `BaseDataset` that inherits from `torch.utils.data.Dataset` and contains functions to retrieve, manipulate and cache the downloaded dataset.
It also contains an implementation of a `QM9Dataset` that  provides functions that control from which remote location the dataset is retrieved and the transformations to generate an opimized npy file. It also implements the `__getitem__` and `__len__` calls necessary for the   `torch.utils.data.Dataset` class.

## Todos
  - [x] outline package structure
  - [x] define dataset module
 
## Status
- [x] Ready to go

### CODE REVIEW

File: modelforge/tests/test_dataset.py [hunk 1]
```diff
@@ -1,17 +1,116 @@
...
+    assert np.array_equal(padded_data[-1][-1], np.array([-1, -1, -1]))
+
+
 @pytest.mark.parametrize("dataset", [QM9Dataset])
 def test_download_dataset(dataset):
-    print(dataset.name)
-    dataset("tmp.hdf5")
+    """Test the downloading and batching of the dataset."""
+    d = dataset(dataset_name="tmp.hdf5", test_data=True)
+    print(d.dataset_name)
```
Comment (chrisiacovella) [inline modelforge/tests/test_dataset.py:98]: I'm guessing this print statement was leftover from testing.  Should this instead be an assert statement with the expected name?

File: modelforge/dataset/qm9.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import gdown
+import h5py
+import numpy as np
+import torch
+from loguru import logger
+from .dataset import BaseDataset
+
+
+class PlainQM9Dataset(torch.utils.data.Dataset):
+    """
```
Comment (chrisiacovella) [inline modelforge/dataset/qm9.py:13]: What is the purposed of this class compared to !m9Dataset?

File: modelforge/dataset/qm9.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        """
+        Initialize the QM9Dataset class.
+
+        Parameters:
+        -----------
+        dataset_name : str
+            Name of the dataset, default is "QM9".
+        load_in_memory : bool
+            Flag to determine if the dataset should be loaded into memory, default is True.
+        test_data : bool
```
Comment (chrisiacovella) [inline modelforge/dataset/qm9.py:95]: I was slightly confused by this at first.  Might want to just mark this something like "for development purposes."

**Review by chrisiacovella — APPROVED** [review]

File: modelforge/dataset/qm9.py (hunk 2 above)
Comment (wiederm) [reply modelforge/dataset/qm9.py:13]: I will remove this class.

File: modelforge/dataset/qm9.py (hunk 3 above)
Comment (wiederm) [reply modelforge/dataset/qm9.py:95]: agreed. changed the variable name.

File: modelforge/tests/test_dataset.py (hunk 1 above)
Comment (wiederm) [reply modelforge/tests/test_dataset.py:98]: thanks! added the assert statment.

**Pull request merged**